Misc/NEWS.d/next/Core_and_Builtins/2025-03-03-20-02-45.gh-issue-130775.fEM6T-.rst

```diff
@@ -0,0 +1,2 @@
+Validate negative :mod:`ast` locations, we only allow ``-1`` as a special
+value.
```

Comment:
Since it's a user-facing change, I think we should also have a What's New entry, and a `versionchanged` in `ast.*`. Users may be surprised. I also don't know whether you want to backport the change or not.

Reply:
yes, I would agree that this is a bug fix for a crash in compile, not a new feature.

Reply:
It's a new error for ast.parse though.

Reply:
Hm, looks like `ast.parse` is not affected:

```python
>>> import ast
... 
... tree = ast.Module(body=[
...     ast.Import(names=[ast.alias(name='traceback', lineno=0, col_offset=0)], lineno=0, \
col_offset=-2)
... ], type_ignores=[])
... 
>>> ast.parse(tree_
... 
KeyboardInterrupt
>>> ast.parse(tree)
Module(body=[Import(names=[alias(name='traceback', asname=None)])], type_ignores=[])
>>> compile(tree, "<string>", "exec")
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    compile(tree, "<string>", "exec")
    ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: AST node has invalid location (0, 0, -2, -2)
```

Reply:
Yes, this works:

```python
>>> ast.parse(tree, optimize=2)
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    ast.parse(tree, optimize=2)
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/Users/sobolev/Desktop/cpython/Lib/ast.py", line 53, in parse
    return compile(source, filename, mode, flags,
                   _feature_version=feature_version, optimize=optimize)
ValueError: AST node has invalid location (0, 0, -2, -2)
```